### ISSUE TITLE
Cannot use GHC 7.10.1-RC2 with xmonad (or any environment)

### ISSUE THREAD

**Issue by MP2E** [NixOS/nixpkgs#6015]
I understand GHC 7.10.1 isn't production-ready yet, but I think the issue is on our end.

I get this error message when attempting to use GHC 7.10.1-RC2 to rebuild my system, after every library individually compiled and installed to store : 
http://lpaste.net/119368

This is what my configs look like:
https://github.com/MP2E/nix-projects/commit/dd0048842f9f7d7ce7a953ae056a88cd86702460

Very similar errors happen if I try to install haskellEnv(also in that commit) via nix-env. Both of these environments work fine under GHC 7.8.4

**Comment by peti** [comment]
These errors are caused by GHC's wonderful "non-determinism" feature. My guess is that you mixed up locally built binaries with ones downloaded from Hydra. @Fuuzetsu might be interested in asking you some questions because he's working on fixing that issue in GHC. Anyway, the only way to remedy this is issue is to run `nix-store --delete <GHC>` where `GHC` is the store path of your local 7.10.1-rc2 compiler. This will delete GHC and *all* package it has built from your system. Now re-build your your environment and hope for the best. :-)

Anyway, this issue is not related to Nix; it's a "feature" of GHC and this issue occurs randomly with all versions of GHC, unfortunately.

**Comment by Fuuzetsu** [comment]
I am indeed interested in such cases but this one is something we can not fix easily: before GHC 7.10.x, the non-determinism occured often. This means that your GHC 7.10.x was bootstrapped with a compiler that produced ‘unexpected’ ABI hashes: it is just the 7.8.4 (and earlier) problem applied to GHC 7.10 core libs. So as peti says, the only thing you can do is get rid of that build and try again, preferably getting a binary from Hydra or whatever cache you use. Once you obtain a GHC 7.10.x with ABIs expected by rest of packages, it should be smooth sailing: if you have ABI problems compiling stuff with GHC 7.10 itself then that is of great interest to me and I would love to hear about it, get the contents of the relevant store paths and any further info.

I don't really see a way out of this particullar issue except trying until you get good hashes for your 7.10. We could do something like compile our 7.10 with a binary 7.10 but that seems a bit silly.

The GHC Trac number is 4012 if you're interested in looking at the problem.

**Comment by MP2E** [comment]
Thanks for the feedback, your explanations make a lot of sense, however I still cannot get any environments to build. It keeps giving the same error.

This is what I tried:

1st attempt
- I used 'ls /nix/store/*ghc-7.10*' to get every ghc 7.10.1 path from the store(I found 2)
- I used nix-store --delete to delete both of the store paths and a few other paths as well
- Rebuild, note that build cache is grabbed this time

2nd attempt
- delete the 2 ghc 7.10 store paths using the above steps again
- Rebuild, pass option 'use-binary-caches false'

3rd attempt
- Uninstall current GHC 7.8.4 environment from user, remove all traces of haskell and xmonad from configuration.nix, rebuild
- Reboot and run 'sudo nix-collect-garbage -d'
- Check /nix/store is empty of Haskell packages via 'ls /nix/store/*haskell* && ls /nix/store/*ghc*'
- Rebuild, noting that build cache is grabbed again

4th attempt
- Same as 3rd attempt except passed option 'use-binary-caches false'

I'm not sure if this is still the problem described, so I'll reopen this issue just to be safe.

**Comment by peti** [comment]
When `nix-store --delete` is run, does it warn you that some paths could not be deleted because they're still "alive"?

**Comment by MP2E** [comment]
Nope, everything is deleted as expected.
````
cray:~$ ls /nix/store/*ghc-7.10*                                                                                                                                                                                                              ~
/nix/store/84k492pz1ximh1w8kp0hw6kc7wldp88d-ghc-7.10.0.20140123.drv  /nix/store/izycn3w04g7c3pjxzmahg0nw47wai3ky-ghc-7.10.0.20140123.lock
/nix/store/aj7whrh49gz4ckw655n4v0ykh1nzkbfl-ghc-7.10.0.20140123.drv  /nix/store/jsrzcbdiany7r3fw84pl7szyidghb54i-ghc-7.10.0.20150123-src.tar.xz.drv

/nix/store/gfma5633jzysxdfhm35cq3qdrvs9ixf5-ghc-7.10.0.20140123:
bin/  lib/  share/

/nix/store/izycn3w04g7c3pjxzmahg0nw47wai3ky-ghc-7.10.0.20140123:
bin/  lib/  share/
cray:~$ nix-store --delete /nix/store/gfma5633jzysxdfhm35cq3qdrvs9ixf5-ghc-7.10.0.20140123 /nix/store/izycn3w04g7c3pjxzmahg0nw47wai3ky-ghc-7.10.0.20140123                                                                                    ~
finding garbage collector roots...
removing stale link from ‘/nix/var/nix/gcroots/auto/p6gys4x56v4gp0373ajcc53glaivg44m’ to ‘/tmp/nix-build.ev1mC1/derivation’
removing stale link from ‘/nix/var/nix/gcroots/auto/qypaf756fmpfyjhh0px0s6wjdrnd50pq’ to ‘/tmp/nixos-rebuild.2j9QMT/nix’
removing stale link from ‘/nix/var/nix/gcroots/auto/m17j2pniy89g3nyr02yddcn6gab1wr9h’ to ‘/tmp/nixos-rebuild.JCTLIF/nix’
removing stale link from ‘/nix/var/nix/gcroots/auto/6sgffhd89d42hp6f33c4x7m52di9hy2r’ to ‘/tmp/nix-build.KYaqLO/derivation’
deleting ‘/nix/store/01kh1blyp3rk40c23bdli4ysg54gs8mz-haskell-test-framework-hunit-0.3.0.1’
deleting ‘/nix/store/lxal05q3rcmg6ika7cskg4a290i1jcx6-haskell-QuickCheck-2.7.6’
deleting ‘/nix/store/12dcchcjw7h9ypq5nvd2l0x85hgk1r73-haskell-tf-random-0.5’
deleting ‘/nix/store/azka92w6a9318z626whv9886kr9xddzn-haskell-xmonad-extras-0.12’
deleting ‘/nix/store/7zslh5pb1vx7837wav4q9n0dwj9qbxbs-haskell-xmonad-contrib-0.12’
deleting ‘/nix/store/w115aplshayjls5ikm484x2y2pb9rrbs-haskell-test-framework-0.8.1.1’
deleting ‘/nix/store/0g5p9qaf1wxz4fmfbm9raym01b3w38ky-haskell-random-1.1’
deleting ‘/nix/store/0cswlz5dqdficy5iss7j7vwa6anganfy-unit-polkit.service’
deleting ‘/nix/store/rpy6m15pw18s6f8nvm3kch2ql45yfp95-dbus-conf’
deleting ‘/nix/store/6iwfhsmgzyakiwmpsvgs9518h4s31zgp-unit-cron.service’
deleting ‘/nix/store/6pls8fgqjkihmzhwh62c1cbc98xfdl5m-unit-script’
deleting ‘/nix/store/zzh6kjsd8llshghlchzb9yjg5gn679p3-system-crontab’
deleting ‘/nix/store/8dihg1g1pni0jxkwfz9nj36jr2ik78n1-system-path’
deleting ‘/nix/store/21h5n76v5c54c1ghb7dn1r8jg5j28rkz-haskell-xmonad-0.12’
deleting ‘/nix/store/a82rjl6ddrp4p8n1xy2yapj07xarbmyy-haskell-parsec-3.1.8’
deleting ‘/nix/store/qh9rxrfzhh3cmrnc69jc2kxmw27ic47h-haskell-regex-posix-0.95.2’
deleting ‘/nix/store/wji41cqdfwmjjk2mzrjkkyyd46cgkrw1-haskell-regex-base-0.93.2’
deleting ‘/nix/store/2mxfbkablpz2qllbk4jd7xnpi5m0rqzg-haskell-mtl-2.2.1’
deleting ‘/nix/store/v0qni3g1nvf5p1ncx7hpjiaf6bn6qjdf-haskell-X11-xft-0.3.1’
deleting ‘/nix/store/pxqzb6n3fym54i1lxy65zl6gz749psjf-haskell-X11-1.6.1.2’
deleting ‘/nix/store/xy6450x8k3b7hf75gn9dxj2335ngis7d-haskell-data-default-0.5.3’
deleting ‘/nix/store/4v79qspy5ais9mzcmfj8il9v549pam15-haskell-data-default-instances-base-0.0.1’
deleting ‘/nix/store/asrbyzhzzkd8lj17bxdc4pxdz7wrz95n-haskell-data-default-instances-old-locale-0.0.1’
deleting ‘/nix/store/f1b70vaph6wp369z0a5hyr6jik6bgy7i-haskell-data-default-instances-containers-0.0.1’
deleting ‘/nix/store/k5bifnjh1ckxnisy94h6jbp46cn4c26j-haskell-data-default-instances-dlist-0.0.1’
deleting ‘/nix/store/33ki4x4z4x9kir23ha6r3rn49lkanq5y-haskell-data-default-class-0.0.1’
deleting ‘/nix/store/6r490fmd4nqdf7nzg5wzwrl38l6sqcyc-haskell-primitive-0.5.4.0’
deleting ‘/nix/store/7gk7z8pk6bzll7q6k8p36b7lpi4f1li8-haskell-HUnit-1.2.5.2’
deleting ‘/nix/store/7xd1aigijk9gh9kw74ic61w98a7dfz9s-hscolour-1.20.3’
deleting ‘/nix/store/whh8ws1pb14kal2kay2f81l87gm4rzys-haskell-xml-1.3.13’
deleting ‘/nix/store/9qala5mz8hs13wy00b8xkl7jipb7pzjh-haskell-text-1.2.0.4’
deleting ‘/nix/store/c64kanyinfqcygswysw5zfqp867r728g-haskell-utf8-string-0.3.8’
deleting ‘/nix/store/lj7d3kqqagwjqx9xgmqfqwr9bibhin6v-haskell-old-time-1.1.0.3’
deleting ‘/nix/store/dp8rsx2siysav32v52kfy4zrmpv50cd2-haskell-old-locale-1.0.0.7’
deleting ‘/nix/store/fshszvg58331ywiyha15nj5nxpb3c8q9-haskell-ansi-wl-pprint-0.6.7.1’
deleting ‘/nix/store/fyar6f509binp52cq7s9nxmz3yhw67bf-haskell-setlocale-1.0.0.1’
deleting ‘/nix/store/kn7a67j3w9b94091h9z84sdh8l71gnn5-jailbreak-cabal-1.2’
deleting ‘/nix/store/lwcr4h8jbyv67046p4wzzi3vgk1jg0x8-haskell-split-0.2.2’
deleting ‘/nix/store/nkgq6b1z3hfrhyaxrnj7ldfcmv2pkdrb-haskell-ansi-terminal-0.6.2.1’
deleting ‘/nix/store/q1k14164nl5rf6rd9kzabqxpr70vh16f-haskell-dlist-0.7.1’
deleting ‘/nix/store/r75hgp6k3mjcdmcqc5v9x6rlsy7jnsbs-haskell-extensible-exceptions-0.1.1.4’
deleting ‘/nix/store/zpnlvgicgbk56zvzyraba99x0sgwi64x-haskell-hostname-1.0’
deleting ‘/nix/store/gfma5633jzysxdfhm35cq3qdrvs9ixf5-ghc-7.10.0.20140123’
deleting ‘/nix/store/izycn3w04g7c3pjxzmahg0nw47wai3ky-ghc-7.10.0.20140123’
deleting ‘/nix/store/trash’
deleting unused links...
note: currently hard linking saves 385.65 MiB
44 store paths deleted, 1106.53 MiB freed
````

**Comment by peti** [comment]
Interesting. I'm not sure how to explain that phenomenon. The `use-binary-caches false` flag should guarantee that your system is consistent. As far as I can tell, it should be impossible for that error to occur when you've build the Haskell compiler and all packages locally.

I'll have to think about this.

**Comment by peti** [comment]
Okay. The problem is that some well-known troublemaker has broken the GHC 7.10.1 derivation in https://github.com/NixOS/nixpkgs/commit/a9b774204b6570593635cfb9f46b20cc2f9246f1: Nix and the compiler disagreed on its version number, and as a result the wrapper script failed to symlink the compiler's base libraries into the generated environment. That issue has been fixed in 86753ed.

The current `master` branch should work fine for you.

**Comment by MP2E** [comment]
@peti Sorry about that...
